File: Creational/Lab5/Python/house.py
from real_estate import *

class House(RealEstate):
    def __init__(self, address="Dungeon", description="tra ta ta", area=50, price=1000, status="New"):
        super().__init__(address, description, area, price, status)
        self.garden = False
        self.pool = False
        self.gym = False
        self.heating_type = None

    def __str__(self):
        return f"House:\n{super().__str__()}"

    def rent(self, customer, start_date, end_date):
        if self.is_rentable:
            print(f"Customer {customer.name_and_surname} rented house on address {self.address} from {start_date} to {end_date}.")

    def sell(self, customer):
        if self.is_sellable:
            print(f"Customer {customer.name_and_surname} bought house on address {self.address}.")

    def buy(self, customer):
        if self.is_buyable:
            print(f"Customer {customer.name_and_surname} sold house on address {self.address}.")


    def clone(self):
        clone = copy.deepcopy(self)
        clone.realtor = None if clone.realtor is None else Realtor( self.realtor.name_and_surname, self.realtor.phone, self.realtor.email)
        return clone

    def shallow_copy(self):
        return copy.copy(self)
        
    def rent(self, customer, start_date, end_date):
        if self.isRentable:
            print(f"House rented to {customer.name_and_surname} from {start_date} to {end_date}")

    def sell(self, customer):
        if self.isSellable:
            print(f"House sold to {customer.name_and_surname}")

File: Creational/Lab6/C#/House.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_
{
    public class House : RealEstate
    {

        public bool Garden { get; set; }
        public bool Pool { get; set; }
        public bool Gym { get; set; }
        public string HeatingType { get; set; }
        public House():base("Dungeon", "tra ta ta", 50, 1000, "New"){
            IsRentable = true;
            IsSellable = false;
        }
        public House(string address, string description, double area, double price, string status): 
            base(address, description, area, price, status) 
        {   
            IsRentable = true;
            IsSellable = false;
        }

        public override string ToString()
        {
            return "House:\n" + base.ToString();
        }

        public override void Rent(Customer customer, DateTime startDate, DateTime endDate){
            if(IsRentable)
                Console.WriteLine($"Customer {customer.NameAndSurname} rented house on address {Address} from {startDate} to {endDate}.");
        }

        public override void Sell(Customer customer){
            if(IsSellable)
                Console.WriteLine($"Customer {customer.NameAndSurname} bought house on address {Address}.");
        }
        public override void Buy(Customer customer){
            if(IsBuyable)
                Console.WriteLine($"Customer {customer.NameAndSurname} sold house on address {Address}.");
        }
        public override House Clone()
        {
            var clone = (House)MemberwiseClone();
            clone.Realtor = Realtor == null ? Realtor : new Realtor(Realtor);
            return clone;
        }

        public override House ShallowCopy()
        {
            return (House)MemberwiseClone();
        }

        public override void Reset()
        {
            base.Reset();
            HeatingType = "";
            Gym = false;
            Garden = false;
            Pool = false;
        }
    }
}
